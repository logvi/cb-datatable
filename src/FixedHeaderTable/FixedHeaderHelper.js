class FixedHeaderHelper {
  constructor({
      tableBodyContainer,
      tableBody,
      tableHeaderContainer,
      tableHeader
  }) {
      this.tableBodyContainer = tableBodyContainer;
      this.tableBody = tableBody;
      this.tableHeaderContainer = tableHeaderContainer;
      this.tableHeader = tableHeader;

      this.tableBodyContainer && this.tableBodyContainer.addEventListener('scroll', this.tableBodyScrollHandler);
      this.adjustColumns();
  }

  tableBodyScrollHandler = ({target}) => {
      const {scrollLeft} = target;
      this.tableHeaderContainer.style.transform = 'translate3d(' + (-1 * scrollLeft) + 'px, 0, 0)';
  };

  adjustColumns() {
    if (!this.tableBody || !this.tableBody.rows.length) return;
    const tableBodyCells = this.tableBody.rows[0].cells;

    for (let i = 0; i < tableBodyCells.length; i++) {
      const tableHeaderCell = this.tableHeader.rows[0].cells[i];
      const tableBodyCellWidth = tableBodyCells[i].getBoundingClientRect().width;
      const tableHeaderCellWidth = tableHeaderCell.getBoundingClientRect().width;
      if (tableBodyCellWidth === tableHeaderCellWidth) continue;
      const newWidth = Math.max(tableBodyCellWidth, tableHeaderCellWidth);
      this.tableHeader.rows[0].cells[i].style['min-width'] = newWidth + 'px';
      tableBodyCells[i].style['min-width'] = newWidth + 'px';
    }
  }
}

export default FixedHeaderHelper;
