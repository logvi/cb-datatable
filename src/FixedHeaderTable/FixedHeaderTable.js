// @flow
import React from 'react';
import TableHeader from '../TableHeader';
import TableBody from '../TableBody';
import FixedHeaderHelper from './FixedHeaderHelper';
import type {TableProps} from '../types';

class FixedHeaderTable extends React.Component<TableProps> {
  props: TableProps;
  fixedHeaderHelper: FixedHeaderHelper;
  tableContainer: HTMLElement;
  tableBodyContainer: ?HTMLElement;
  tableBody: ?HTMLElement;
  tableHeaderContainer: ?HTMLElement;
  tableHeader: ?HTMLElement;

  componentDidMount() {
    this.fixedHeaderHelper = new FixedHeaderHelper({
      tableContainer: this.tableContainer,
      tableBodyContainer: this.tableBodyContainer,
      tableBody: this.tableBody,
      tableHeaderContainer: this.tableHeaderContainer,
      tableHeader: this.tableHeader
    });
  }

  componentDidUpdate() {
    this.fixedHeaderHelper.adjustColumns();
  }

  get Header(): React$Element<*> {
    return (
      <div className="fixed-header-table__header" ref={el => this.tableHeaderContainer = el}>
        <table ref={el => this.tableHeader = el}>
          <TableHeader orderBy={this.props.orderBy} setOrderBy={this.props.setOrderBy} columns={this.props.columns} />
        </table>
      </div>
    )
  }

  get Body(): React$Element<*> {
    return (
      <div className="fixed-header-table__body" ref={el => this.tableBodyContainer = el}>
        <table ref={el => this.tableBody = el}>
          <TableBody
            BodyRowRenderer={this.props.BodyRowRenderer}
            columns={this.props.columns}
            data={this.props.displayData}
            onRowClick={this.props.onRowClick}
          />
        </table>
      </div>
    )
  }

  get Loader(): ?React$Element<*> {
    if (!this.props.loading) return;
    return this.props.Loader;
  }

  renderLayout() {
    return (
      <div className="fixed-header-table">
        {this.Header}
        {this.Loader}
        {this.Body}
      </div>
    )
  }

  render() {
    return (
      <span className="cb-datatable__table" ref={el => this.tableContainer = el}>
        {this.props.renderTableLayout ?
        this.props.renderTableLayout({Body: this.Body, Header: this.Header, Loader: this.Loader})
        : this.renderLayout()}
      </span>
    )
  }
}

export default FixedHeaderTable;
