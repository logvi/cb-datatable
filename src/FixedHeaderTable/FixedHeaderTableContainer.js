import DataTableContainer from '../DataTableContainer';
import FixedHeaderTable from './FixedHeaderTable';

class FixedHeaderTableContainer extends DataTableContainer {
  static defaultProps = {
    ...DataTableContainer.defaultProps,
    TableComponent: FixedHeaderTable
  };
}

export default FixedHeaderTableContainer;
