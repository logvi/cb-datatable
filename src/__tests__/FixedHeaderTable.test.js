import FixedHeaderTable from '../FixedHeaderTable';
import GeneralTableTests from './GeneralTableTests';

describe('FixedHeaderTable', GeneralTableTests(FixedHeaderTable));
