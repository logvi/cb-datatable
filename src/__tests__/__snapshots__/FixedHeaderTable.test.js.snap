// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedHeaderTable Columns visibility. Column 1 should be hidden 1`] = `
<div
  className="cb-datatable"
  onClick={undefined}
>
  <div
    className="fixed-header-table"
  >
    <div
      className="fixed-header-table__header"
    >
      <table>
        <thead>
          <tr>
            <th
              className="sortable"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                column0
              </div>
            </th>
            <th
              className="sortable"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                column2
              </div>
            </th>
            <th
              className="sortable"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                column3
              </div>
            </th>
            <th
              className="sortable"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                column4
              </div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div
      className="fixed-header-table__body"
    >
      <table>
        <tbody>
          <tr
            onClick={null}
          >
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column0 0
              </div>
            </td>
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column2 0
              </div>
            </td>
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column3 0
              </div>
            </td>
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column4 0
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`FixedHeaderTable orderBy asc Snapshot: should apply sort class and icon 1`] = `
<div
  className="cb-datatable"
  onClick={undefined}
>
  <div
    className="fixed-header-table"
  >
    <div
      className="fixed-header-table__header"
    >
      <table>
        <thead>
          <tr>
            <th
              className="sortable sort-asc"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                <div
                  className="th-content__label"
                >
                  column 0
                </div>
                <span
                  className="icon-sort icon-sort-asc"
                >
                   ▲
                </span>
              </div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div
      className="fixed-header-table__body"
    >
      <table>
        <tbody>
          <tr
            onClick={null}
          >
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column0 0
              </div>
            </td>
          </tr>
          <tr
            onClick={null}
          >
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column0 1
              </div>
            </td>
          </tr>
          <tr
            onClick={null}
          >
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column0 2
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`FixedHeaderTable orderBy desc Snapshot: should apply sort class and icon 1`] = `
<div
  className="cb-datatable"
  onClick={undefined}
>
  <div
    className="fixed-header-table"
  >
    <div
      className="fixed-header-table__header"
    >
      <table>
        <thead>
          <tr>
            <th
              className="sortable sort-desc"
              onClick={[Function]}
            >
              <div
                className="th-content"
              >
                <div
                  className="th-content__label"
                >
                  column 0
                </div>
                <span
                  className="icon-sort icon-sort-desc"
                >
                   ▼
                </span>
              </div>
            </th>
          </tr>
        </thead>
      </table>
    </div>
    <div
      className="fixed-header-table__body"
    >
      <table>
        <tbody>
          <tr
            onClick={null}
          >
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column0 2
              </div>
            </td>
          </tr>
          <tr
            onClick={null}
          >
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column0 1
              </div>
            </td>
          </tr>
          <tr
            onClick={null}
          >
            <td
              className={null}
            >
              <div
                className="table-cell-content"
              >
                column0 0
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`FixedHeaderTable should apply renderLayout prop, it displays only table 1`] = `
<div
  className="fixed-header-table"
>
  <div
    className="fixed-header-table__header"
  >
    <table>
      <thead>
        <tr>
          <th
            className="sortable"
            onClick={[Function]}
          >
            <div
              className="th-content"
            >
              column 0
            </div>
          </th>
        </tr>
      </thead>
    </table>
  </div>
  <div
    className="fixed-header-table__body"
  >
    <table>
      <tbody>
        <tr
          onClick={null}
        >
          <td
            className={null}
          >
            <div
              className="table-cell-content"
            >
              column0 0
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
